comms/src/peer_manager/peer.rs

```diff
@@ -39,13 +43,76 @@ bitflags! {
 }
 
 /// Peer connection statistics
-#[derive(Debug, Clone, Default, Deserialize, Serialize, PartialOrd, PartialEq)]
-pub struct PeerConnectionStats {
-    /// The last time that a successful connection was made, None if this has never happened
-    pub connected_at: Option<NaiveDateTime>,
-    /// The time that the last connection attempt failed. If the last connection attempt succeeded
-    /// this is None
-    pub last_connect_failed_at: Option<NaiveDateTime>,
+#[derive(Debug, Clone, Deserialize, Serialize, PartialOrd, PartialEq)]
+pub enum LastConnectionState {
+    /// This node has never connected to this peer
+    Never,
+    /// The last connection attempt was successful
+    Succeeded(NaiveDateTime),
+    /// The last connection attempt failed.
+    Failed {
+        /// Timestamp of the last failed attempt
+        failed_at: NaiveDateTime,
+        /// Number of failed attempts in a row
+        num_attempts: usize,
+    },
+}
+
+impl LastConnectionState {
+    /// Returns the number of failed attempts. 0 is returned if the `LastConnectionState` is not `Failed`
+    pub fn failed_attempts(&self) -> usize {
+        match self {
+            LastConnectionState::Failed { num_attempts, .. } => *num_attempts,
+            _ => 0,
+        }
+    }
+
+    /// Returns the date time (UTC) since the last failed connection occurred. None is returned if the
+    /// `LastConnectionState` is not `Failed`
+    pub fn failed_at(&self) -> Option<&NaiveDateTime> {
+        match self {
+            LastConnectionState::Failed { failed_at, .. } => Some(failed_at),
+            _ => None,
+        }
+    }
+
+    /// Returns the Duration since the last failed connection occurred. None is returned if the `LastConnectionState` is
+    /// not `Failed`
+    pub fn failed_since_now(&self) -> Option<Duration> {
```

Comment:
This is a weird name. time_since_last_failure?

Reply:
One day I'll figure out this naming thing 😆